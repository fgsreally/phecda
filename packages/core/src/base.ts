import { Empty, Unmount } from './decorators'
import { getTag } from './helper'
import { Events } from './types'

@Empty
export abstract class Base {
  private readonly __UNMOUNT_SYMBOL__: (() => void)[] = []
  private readonly __PROMISE_SYMBOL__: Promise<void> // for init

  abstract emitter: any
  constructor() {

  }

  get tag(): PropertyKey {
    return getTag(this)
  }

  then(cb: () => void, reject?: (e: any) => void) {
    return this.__PROMISE_SYMBOL__.then(cb, reject)
  }

  on<Key extends keyof Events>(type: Key, handler: (arg: Events[Key]) => void): void {
    this.emitter.on(type, handler)
    this.onUnmount(() => this.emitter.off(type, handler))
  }

  emit<Key extends keyof Events>(type: Key, param: Events[Key]) {
    this.emitter.emit(type, param)
  }

  off<Key extends keyof Events>(type: Key, handler?: (arg: Events[Key]) => void): void {
    this.emitter.off(type, handler)
  }

  protected onUnmount(cb: () => void) {
    this.__UNMOUNT_SYMBOL__.push(cb)
  }

  @Unmount
  // @ts-expect-error for internal
  private _unmount() {
    return Promise.all(this.__UNMOUNT_SYMBOL__.map(fn => fn()))
  }
}
