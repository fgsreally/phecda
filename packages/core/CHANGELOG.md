# phecda-core

## 2.0.1-alpha.0

### Patch Changes

- 1dd0831: isPhecda should return true when class is decorated by any phecda-core decorator

## 2.0.0

### Major Changes

- 78cb57a: refactor namespace structor to avoid namespace population

### Patch Changes

- d6d2146: Global will mount class in init handler
- 049c138: plainToClass support option rule
- a701f34: Inject won't throw error if not Provide
- 25cf638: fix Global when extend
- eec80a6: no essential to use Global before Tag
- 64c2f70: Rule support function info
- 8fe5ced: export all presets

## 2.0.0-alpha.7

### Patch Changes

- a701f34: Inject won't throw error if not Provide

## 2.0.0-alpha.6

### Patch Changes

- 25cf638: fix Global when extend

## 2.0.0-alpha.5

### Patch Changes

- eec80a6: no essential to use Global before Tag

## 2.0.0-alpha.4

### Patch Changes

- d6d2146: Global will mount class in init handler

## 2.0.0-alpha.3

### Patch Changes

- 049c138: plainToClass support option rule

## 2.0.0-alpha.2

### Patch Changes

- 8fe5ced: export all presets

## 2.0.0-alpha.1

### Patch Changes

- 64c2f70: Rule support function info

## 2.0.0-alpha.0

### Major Changes

- 78cb57a: refactor namespace structor to avoid namespace population

## 1.7.0

### Minor Changes

- 8022370: add Effect decorator

## 1.6.0

### Minor Changes

- 2daabb8: add Provide and Inject

## 1.5.0

### Minor Changes

- 538d86f: add Bind and getBind

## 1.4.1

### Patch Changes

- 4b67e98: rename P to Expose

## 1.4.0

### Minor Changes

- 263a6a7: add Assign to assign value to instance

## 1.3.1

### Patch Changes

- caab8b6: remove type prefix

## 1.3.0

### Minor Changes

- f13c1e9: remove **TAG** to prototype, avoid pollution in namespace

## 1.2.1

### Patch Changes

- 4273f22: external preset and add iife format bundle

## 1.2.0

### Minor Changes

- 9b91e2f: Storage is an 'inject' decorator like Watcher at now

## 1.1.1

### Patch Changes

- add types limit to Watcher, must be declared in a type file
- 0b46788: add option to watcher

## 1.1.0

### Minor Changes

- 589aaad: Clear and Ignore will work in each situation

## 1.0.8

### Patch Changes

- replace window with global,get with p

## 1.0.7

### Patch Changes

- a60ca7e: plainToClass can only validate or transform ,not both

## 1.0.6

### Patch Changes

- add isPhecda

## 1.0.5

### Patch Changes

- add window decorator

## 1.0.4

### Patch Changes

- add getTag utils

## 1.0.3

### Patch Changes

- 24e2f22: move all decorator to core and other pkg only need to inject property

## 1.0.2

### Patch Changes

- b119020: add more pipe and rule to core

## 1.0.1

### Patch Changes

- refactor:remove phecda-form code and part of phecda-core code to phecda-vue to make sure all content that is related to vue is in phecda-vue;remove package phecda-form
