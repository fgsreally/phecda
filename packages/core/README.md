# phecda-core
provide base function and abstract limit to other phecda module

