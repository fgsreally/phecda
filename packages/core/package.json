{
  "name": "phecda-core",
  "version": "3.0.0-alpha.7",
  "description": "provide base function and abstract limit to other phecda module ",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "scripts": {
    "dev": "tsup --watch",
    "build": "tsup"
  },
  "keywords": [],
  "author": "",
  "files": [
    "dist"
  ],
  "license": "MIT",
  "devDependencies": {
    "tsup": "^6.5.0"
  }
}