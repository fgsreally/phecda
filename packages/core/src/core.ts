import type { Phecda, PhecdaHandler } from './types'
import { mergeOptions } from './utils'

export function isPhecda(target: any) {
  return target && !!target.prototype._namespace
}
// 一个类挂载第一个phecda装饰器时，会创建对应的，类似元数据的东西
export function init(target: Phecda) {
  if (!target._namespace) {
    target._namespace = {

      /**
         * @Init 事件,
         * 引入Model时就会执行
        */
      __INIT_EVENT__: new Set(),
      /**
         * 暴露的变量，
         * 只要属性上存在至少一个装饰器，该属性就会被捕捉到
        */
      __EXPOSE_VAR__: new Set(),
      /**
         * @Ignore 绑定的属性，
         * 某属性即使被捕捉，可被强行忽略，优先级最高
        */
      __IGNORE_VAR__: new Set(),
      /**
         * 存在状态的变量
         * @deprecated
        */
      __STATE_VAR__: new Set(),
      /**
         * 状态变量的处理器
        */
      __STATE_HANDLER__: new Map(),
      /**
         * 状态变量的共有状态
        */
      __STATE_NAMESPACE__: new Map(),
    }
  }
}

export function regisInitEvent(target: Phecda, key: string) {
  init(target)
  target._namespace.__INIT_EVENT__.add(key)
}

export function getInitEvent(target: Phecda) {
  init(target)
  return [...target._namespace.__INIT_EVENT__] as string[]
}

export function setModalVar(target: Phecda, key: PropertyKey) {
  init(target)
  target._namespace.__STATE_VAR__.add(key)
  // 绑定状态的值，均属于暴露的值
  setExposeKey(target, key)
}

export function setExposeKey(target: Phecda, key: PropertyKey) {
  init(target)
  target._namespace.__EXPOSE_VAR__.add(key)
}

export function setIgnoreKey(target: Phecda, key: PropertyKey) {
  init(target)
  target._namespace.__IGNORE_VAR__.add(key)
}
// 存在状态的属性
export function getModelState(target: Phecda) {
  init(target)
  return [...target._namespace.__STATE_VAR__] as string[]
}
// 暴露的属性
// 存在状态必然暴露，反之未必，但expose可以被ignore，前者不行
// 一般而言用这个就行，某些特定情况，可用前一种
export function getExposeKey(target: Phecda) {
  init(target)

  return [...target._namespace.__EXPOSE_VAR__].filter(item => !target._namespace.__IGNORE_VAR__.has(item)) as string[]
}

export function getIgnoreKey(target: Phecda) {
  init(target)
  return [...target._namespace.__IGNORE_VAR__] as string[]
}

export function regisHandler(target: Phecda, key: PropertyKey, handler: PhecdaHandler) {
  init(target)
  if (!target._namespace.__STATE_HANDLER__.has(key))
    target._namespace.__STATE_HANDLER__.set(key, [handler])

  else
    target._namespace.__STATE_HANDLER__.get(key)!.push(handler)
}

export function getHandler(target: Phecda, key: PropertyKey) {
  return target._namespace.__STATE_HANDLER__.get(key) || []
}

export function mergeState(target: Phecda, key: PropertyKey, state: any) {
  const namespace = target._namespace.__STATE_NAMESPACE__
  if (!namespace.has(key))
    namespace.set(key, state)
  else mergeOptions(namespace.get(key)!, state)
}

export function getState(target: Phecda, key: PropertyKey) {
  const namespace = target._namespace.__STATE_NAMESPACE__
  if (namespace)
    return namespace.get(key)
}

// work for init
export function register(instance: Phecda) {
  const stateVars = getExposeKey(instance) as PropertyKey[]

  for (const item of stateVars) {
    const handlers = getHandler(instance, item)

    for (const hanlder of handlers)
      hanlder.init?.(instance)
  }
}

export async function registerAsync(instance: Phecda) {
  const stateVars = getExposeKey(instance) as PropertyKey[]

  for (const item of stateVars) {
    const handlers = getHandler(instance, item)
    for (const hanlder of handlers)
      await hanlder.init?.(instance)
  }
}
