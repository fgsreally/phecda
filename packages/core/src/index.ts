export * from './decorators'
export * from './types'
export * from './helper'
export * from './utils'
export * from './core'
export * from './namespace'
export * from './custom/decorator'
export * from './preset'
