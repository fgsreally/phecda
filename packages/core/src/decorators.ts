import type { to } from './helper'
import { init, regisHandler, setExposeKey, setIgnoreKey, setModalVar } from './core'

export function Init(target: any, key: PropertyKey) {
  setModalVar(target, key)

  regisHandler(target, key, {
    async init(instance: any) {
      instance[key]()
    },
  })
}

export function Rule(rule: RegExp | string | Function | number,
  info: string,
  meta?: any) {
  return (obj: any, key: PropertyKey) => {
    setModalVar(obj, key)
    regisHandler(obj, key, {
      rule,
      info,
      meta,
    })
  }
}

export function Ignore(target: any, key: PropertyKey) {
  setIgnoreKey(target, key)
}

export function Clear(target: any, key: PropertyKey) {
  init(target)

  target._namespace.__INIT_EVENT__.delete(key)
  target._namespace.__EXPOSE_VAR__.delete(key)
  target._namespace.__IGNORE_VAR__.delete(key)
  target._namespace.__STATE_VAR__.delete(key)
  target._namespace.__STATE_HANDLER__.delete(key)
  target._namespace.__STATE_NAMESPACE__.delete(key)
}

export function Err<Fn extends (...args: any) => any>(cb: Fn) {
  return (target: any, key: PropertyKey) => {
    setModalVar(target, key)
    regisHandler(target, key, {
      error: cb,
    })
  }
}

export function Expose(target: any, key: PropertyKey) {
  setExposeKey(target, key)
}

export function Pipe(v: ReturnType<typeof to>) {
  return (obj: any, key: PropertyKey) => {
    setModalVar(obj, key)
    regisHandler(obj, key, {
      async pipe(instance: any) {
        const tasks = v.value
        for (const task of tasks)
          instance[key] = await task(instance[key])
      },
    })
  }
}

export function Tag(tag: string) {
  return (target: any) => {
    target.prototype.__TAG__ = tag
  }
}
// async assign value to instance
export function Assign(cb: (instance?: any) => any) {
  return (target: any) => {
    init(target.prototype)
    setModalVar(target.prototype, '__CLASS')
    regisHandler(target.prototype, '__CLASS', {
      init: async (instance: any) => {
        const value = await cb(instance)
        if (value && typeof value === 'object' && !Array.isArray(value)) {
          for (const i in value)
            instance[i] = value[i]
        }
      },
    })
  }
}

export function Global(target: any) {
  if (!(globalThis as any).__PHECDA__)
    (globalThis as any).__PHECDA__ = {}
  const tag = target.prototype.__TAG__
  if (tag)
    (globalThis as any).__PHECDA__[tag] = target
}
