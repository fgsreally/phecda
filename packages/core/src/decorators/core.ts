import { PHECDA_KEY, SHARE_KEY, init, setExposeKey, setHandler, setIgnoreKey, setState, setStateKey } from '../core'
import type { Phecda } from '../types'

export function Init(proto: any, key: PropertyKey) {
  setStateKey(proto, key)

  setHandler(proto, key, {
    async init(instance: any) {
      return instance[key]()
    },
  })
}

export function Unmount(proto: any, key: PropertyKey) {
  setStateKey(proto, key)

  setHandler(proto, key, {
    async unmount(instance: any) {
      return instance[key]()
    },
  })
}

// bind value
// won't assign
export function Bind(value: any) {
  return (proto: any, k: PropertyKey) => {
    setStateKey(proto, k)
    setState(proto, k, {
      value,
    })
  }
}

export function Ignore(proto: any, key?: PropertyKey) {
  if (!key) {
    proto = proto.prototype
    key = SHARE_KEY
  };
  setIgnoreKey(proto, key)
}

export function Clear(proto: any, key?: PropertyKey) {
  if (!key) {
    proto = proto.prototype
    key = SHARE_KEY
  };
  init(proto);

  (proto as Phecda)[PHECDA_KEY].__CLEAR_KEY.add(key)
}

export function Expose(proto: any, key: PropertyKey) {
  setExposeKey(proto, key)
}

export function Empty(model: any) {
  init(model.prototype)
}

export function Injectable() {
  return (target: any) => Empty(target)
}
