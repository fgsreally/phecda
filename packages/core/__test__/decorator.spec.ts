import { describe, expect, it } from 'vitest'
import { getExposeKey } from '../src/core'
import { Assign, Ignore, P, Pipe, Rule, addDecoToClass, classToValue, plainToClass, registerAsync, to } from '../src/index'
describe('validate&transform', () => {
  class Parent {
    @Ignore
    @Rule('phecda', 'name should be phecda')
    @Pipe(to((name: string) => `${name}1`)
      .to(name => `${name}1`)
      .to(name => `${name}1`))
    name: string

    @P
    get fullname() {
      return `${this.name}-core`
    }

    changeName() {
      this.name = 'phecda-changed'
    }
  }
  it('plainToClass', async () => {
    // false
    const { err } = await plainToClass(Parent, { name: 'phecda11' })
    expect(err.length).toBe(1)

    expect(err[0]).toBe('name should be phecda')

    const { data } = await plainToClass(Parent, { name: 'phecda' })
    expect(data.name).toBe('phecda111')
    expect(data.fullname).toBe('phecda111-core')

    data.changeName()

    expect(data.name).toBe('phecda-changed')
    expect(data.fullname).toBe('phecda-changed-core')
  })

  it('classToValue', async () => {
    const { data } = await plainToClass(Parent, { name: 'phecda' })
    expect(classToValue(data)).toMatchSnapshot()
  })

  it('extend', async () => {
    class Child extends Parent {
      @Rule((str: string) => str.length < 5, 'name should be short')
      name: string
    }
    const { err, data } = await plainToClass(Child, { name: 'phecda11' }, { transform: true, collectError: true })
    expect(err.length).toBe(2)
    expect(err[1]).toBe('name should be short')
    expect(data.name).toBe('phecda11111')
    expect(classToValue(data)).toMatchSnapshot()
  })

  it('use function to add decorator', () => {
    class Any {
      name: string
    }

    addDecoToClass(Any, 'name', P)
    expect(getExposeKey(Any.prototype as any)).toMatchSnapshot()
  })
  it('test Assign', async () => {
    @Assign(() => new Promise(resolve => resolve({ key: 'test2' })))
    class Test {
      key = 'test'
    }
    const instance = new Test() as any
    await registerAsync(instance)
    expect(instance.key).toBe('test2')
  })
})
