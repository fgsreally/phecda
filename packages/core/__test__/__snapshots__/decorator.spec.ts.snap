// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`validate&transform > classToValue 1`] = `
{
  "fullname": "2phecda111-core",
}
`;

exports[`validate&transform > extend 1`] = `
{
  "fullname": "2phecda11111-core",
}
`;

exports[`validate&transform > use function to add decorator 1`] = `
[
  "name",
]
`;
